### ISSUE TITLE
Selenium: TimeoutException in complex mask with many update="@form"

### ISSUE THREAD

**Issue by nettozahler** [primefaces/primefaces#8857]
### Describe the bug

Let's try this (maybe) difficult one. Running a Selenium test on Firefox featuring a complex mask with many update="@form" fails with TimeoutException.

It does not happen always so it might be a timing issue or race condition or concurrency problem. I have seen it with

org.primefaces.selenium.component.InputText
org.primefaces.selenium.component.Calendar
org.primefaces.selenium.component.InputTextarea

but interestingly NOT with org.primefaces.selenium.component.InputNumber

In case you cannot reproduce it, here is some more information what happens as far as I understand it.

The PrimeSelenium Guard class sets this indicator before executing the command:
"pfselenium.xhr = 'somethingJustNotNull';"

In waitUntilAjaxCompletes it checks among other stuff:
pfselenium.xhr == null

In the error case it seems that pfselenium.xhr is not reset to null like this debug output shows:

Guard#ajax; ajaxDebugInfo **before methode.invoke**: document.readyState=complete, !window.jQuery=false, jQuery.active=0, !window.PrimeFaces=false, PrimeFaces.ajax.Queue.isEmpty()=true, PrimeFaces.animationActive=false, !window.pfselenium=false, **pfselenium.xhr=somethingJustNotNull**, pfselenium.anyXhrStarted=null, pfselenium.navigating=false

Guard#ajax; ajaxDebugInfo **in error catch clause**: document.readyState=complete, !window.jQuery=false, jQuery.active=0, !window.PrimeFaces=false, PrimeFaces.ajax.Queue.isEmpty()=true, PrimeFaces.animationActive=false, !window.pfselenium=false, **pfselenium.xhr=somethingJustNotNull**, pfselenium.anyXhrStarted=null, pfselenium.navigating=false

Therefore the condition for "Ajax complete" is never fulfilled and a TimeoutException occurs.

### Reproducer

Will provide a pull request with an appropriate integration test (hopefully) producing the error.

### Expected behavior

Selenium test runs fine without TimeoutException.

### PrimeFaces edition

Community

### PrimeFaces version

11

### Theme

Any

### JSF implementation

MyFaces

### JSF version

2.2, 2.3

### Browser(s)

Firefox 100

**Comment by melloware** [comment]
I will say this definitely looks like a race condition.

But I would also say why would you `update="@form"` in a change request of a single field?  I typically only use `update="@form"` on Submit buttons when I want the whole form and its error messages to re-render.  But AJAX inside change events or keypress you want to be selective with what you update.

**Comment by melloware** [comment]
I will say this definitely looks like a race condition.

But I would also say why would you `update="@form"` in a change request of a single field?  I typically only use update="@form" on Submit buttons when I want the whole form and its error messages to re-render.  But AJAX inside change events or keypress you want to be selective with what you update.

**Comment by melloware** [comment]
Actually whenI just try this simple case...

```xml
        <h:form id="form">
            <p:inputText id="inputtext1" value="#{inputText001.value1}">
                <p:ajax process="@this" update="@form"/>
            </p:inputText>
```

I get a `org.openqa.selenium.StaleElementReferenceException: stale element reference: element is not attached to the page document`

**Comment by melloware** [comment]
Actually whenI just try this simple case...

```xml
        <h:form id="form">
            <p:inputText id="inputtext1" value="#{inputText001.value1}">
                <p:ajax process="@this" update="@form"/>
            </p:inputText>
```

I get a `org.openqa.selenium.StaleElementReferenceException: stale element reference: element is not attached to the page document`

**Comment by nettozahler** [comment]
Yes. I also sometimes had the `org.openqa.selenium.StaleElementReferenceException` when `update="@form"` is present. This forced me to fiddle around with

`new WebDriverWait(driver, waitTime).until(ExpectedConditions.elementToBeClickable(locator));`

and other `ExpectedConditions`. It would be great if PrimeSelenium would handle this automatically :-)  It seems to me that `update="@form"` currently breaks the PrimeSelenium implementation in multiple ways.

**Comment by nettozahler** [comment]
But I would like to emphasize that PrimeSelenium is great and does an outstanding job for me to make my PF application testable with automatic GUI tests. With all the heavy Ajax usage it would be a pain to use plain Selenium API only (believe me, I know what I’m talking about :-) ).

**Comment by nettozahler** [comment]
Motivation for using `update="@form"`: consider having a complex mask with many fields with some elaborated business logic where changing one field forces recalculation of many other fields.

It would be quite cumbersome to have an `update="long list of field IDs"` and also harder to maintain if some fields are removed or added over time.

And it gets even worse when the fields to be updated are rendered under various conditions only. If you do not have a common panel or whatever which contains all these fields so one could do `update="myPanel”` there is a problem.

As far as I know the `update="list of field IDs"` is fix (correct me if I’m wrong).

So we would end up with some ugly

```
<c:if condition=”A”>
  <p:inputnumber update="list of field IDs A"/>
</c:if>
<c:if condition=”B”>
  <p:inputnumber update="list of field IDs B"/>
</c:if>
<c:if condition=”C”>
  <p:inputnumber update="list of field IDs C"/>
</c:if>
```

**Comment by nettozahler** [comment]
Besides all that `update="@form"` is valid PF/JSF API usage  :-)

**Comment by stolp** [comment]
We also have PrimeSelenium based tests in conjunction with Selenide which show a flaky timeout behavior at some defined locations and we pulled our hair wondering what is causing this. Thanks a lot for pointing this problem out!

A quick code sweep revealed, that indeed the flaky parts involve update="@form". We use it sparingly, but it definitely does have its merits, especially regarding code change resilience.

I will try to eliminate the update="@form" in our tests where possible and report back if that resolves our problems.

So currently I can confirm that:

1. update="@form" indeed seems to be causing problems in PrimeSelenium tests
2. update="@form" is a commonly used and valid pattern which IMHO should definitely be supported correctly

@nettozahler We usually work around the ugliness of updating by adding CSS classes indicating that a field is to be updated on form actions.

Example:

```
<p:inputnumber update="@(.toBeUpdated)" />

<p:outputText styleClass="toBeUpdated" />

```
Maybe this works for you too?

**Comment by melloware** [comment]
I agree all around we need to deep dive why @form fails. I am actually surprised we haven't had this reported before.

**Comment by melloware** [comment]
as a note to @stolp comment about using CSS to update specific components you can also use the Observer functionality in PF to specifically target your updates... See: https://www.primefaces.org/showcase/ui/ajax/observer.xhtml

**Comment by melloware** [comment]
OK I fixed InputText and InputTextArea to handle `update="@form"` they were definitely broken.

**Comment by nettozahler** [comment]
> as a note to @stolp comment about using CSS to update specific components you can also use the Observer functionality in PF to specifically target your updates... See: https://www.primefaces.org/showcase/ui/ajax/observer.xhtml

Thank you @stolp and @melloware for your suggestions. When I am back at work I will explore if CSS update or PF Observer functionality suits better for me. I will be back at work on Tuesday due to Monday being public holiday in Germany. Can you point out where the updating by CSS classes is documented?

**Comment by stolp** [comment]
Selectors are documented here: [https://primefaces.github.io/primefaces/11_0_0/#/core/searchexpression?id=primefaces-selectors-pfs](https://primefaces.github.io/primefaces/11_0_0/#/core/searchexpression?id=primefaces-selectors-pfs)

In addition [https://www.primefaces.org/showcase/ui/ajax/selector.xhtml](https://www.primefaces.org/showcase/ui/ajax/selector.xhtml) is showcasing this feature.
